### DISCUSSION TITLE
Add way to change the credentials of a connection pool without recreating the whole pool object (secret rotation)

### DISCUSSION THREAD

**Discussion by jankatins** [psycopg/psycopg#851] [Ideas]
We have the requirement to rotate the PG credentials every x days. Currently this means that we have to restart the service or at least replace the connection pool everywhere.

It would be nice if one could add a small helper thread which would look if the mounted credential have been changed and if so, would update the conninfo in the central connection pool and the pool would then drain the connection and replace them with new ones.

It would be even nicer, if one could give the pool a way to query for the conninfo and the pool would do that from time to time.

If there are already recommended ways: it would be nice to give some recipe in the docs, at least my google skills didn't let anything show up :-(

**Comment by dvarrazzo** [comment]
Hello @jankatins 

the use case is not covered, but I can see its usefulness, thank you for detailing it.

I assume we can add a method like `pool.set_conninfo(conninfo: str="", *, kwargs: dict | None = None)` which would:

- replace `conninfo` and `kwargs` set up in the constructor
- remove every connection currently in the pool
- create an adequate number of connection to fill back the pool using the new conninfo
- upon old connection returned to the pool, they would be closed and replaced with a new connection
- reading the `conninfo` and `kwargs` attributes on the pool can be part of the documented and stable interface.

I'm not convinced to make the pool active with a callback to poll for infrastructure change. I think it's the program's responsibility to push for changes, furthermore providing and documenting a method to call is much easier than accepting a callback (libraries are easier than frameworks).

The work above is relatively simple... except closing the connections currently served when they are returned. It might require more changes than the rest, because at the moment the pool doesn't have a link to the connections that were given away and that therefore must be invalidated. We should either keep a map of them or annotate them in a way that will mark the need to close them on putconn.

All in all I am ok with introducing these changes but I'm not sure about the bandwidth I have, for other works I am doing at the moment and because soonish I would like to wrap up with psycopg 3.2. Are you able to provide a MR? Alternatively can your company fund this work?

**Reply by ratkins** [reply]
We have encountered this exact issue.

Why do you need to remove every connection currently in the pool? Only the new one, and subsequent connections, need the new login credentials. The existing connections are fine.

**Comment by jankatins** [comment]
Ok, most of this looks straight forward. For the "expiring" of old connections:

I see that each connection is "marked" with the pool. One idea would be add a "version" in the pool which would simply increase in case the conninfo is updated. In a new `_mark_connection(self, conn)`, we would both add a the `conn._pool = self` and a `conn._pool_conninfo_version = self._conninfo_version`. In `_check_pool_putconn(conn)` we would check the version, if it is different, we woul set `conn._expire_at = monotonic()-1` and then the normal process would remove it.

Another idea would be simply not do anything: given that the default expiry of connections is about one hour, I would assume that old credentials are valid that long.

Re "callback to poll for changes": one nice thing is that the `ConnectionPool` infra already has all the things to do it, we would just need to supply a function and schedule it. But I guess letting the user do this is also fine.

Would any of these ideas be acceptable? If so I would create a MR for this.

**Comment by dvarrazzo** [comment]
> I see that each connection is "marked" with the pool. One idea would be add a "version" in the pool which would simply increase in case the conninfo is updated

Yes, that's an option but, as we are refactoring this code, I'm considering whether adding a dictionary with the "given" connections (mapping `id(conn) -> conn`) wouldn't be a good refactoring, and to stop adding the pool reference to the connection. The mapping would be adequate to answer the question "does this connection belong to this pool" and I remember in the past someone finding the extra info added to the connection problematic (I think they were trying to use the psycopg 3 pool with psycopg 2 connection, which is not an awful idea).

Maintaining this mapping would be done in the same critical sections where `_pool` is manipulated. At a glance just:

- in `getconn()`:

https://github.com/psycopg/psycopg/blob/558cbe7ef7ef6a6d1f6c65e368aaaf9f9a10108e/psycopg_pool/psycopg_pool/pool_async.py#L296

- in `putconn()`:

https://github.com/psycopg/psycopg/blob/558cbe7ef7ef6a6d1f6c65e368aaaf9f9a10108e/psycopg_pool/psycopg_pool/pool_async.py#L775

 
> Another idea would be simply not do anything: given that the default expiry of connections is about one hour, I would assume that old credentials are valid that long.

I don't know about that, for instance if someone uses this feature because they want to rotate credentials immediately to switch to a different server.

**Reply by ratkins** [reply]
> I don't know about that, for instance if someone uses this feature because they want to rotate credentials immediately to switch to a different server.

Ah, I see your objection. We would be happy with the narrow use case of just rotating passwords without even being able to change the database host. In fact that would be superior for us.

**Comment by jankatins** [comment]
> Maintaining this mapping would be done in the same critical sections where _pool is manipulated.

Sounds good, let me try to build a PR for this.

(might take a week longer or so: I'm currently moving flats)

**Comment by dvarrazzo** [comment]
The feature hasn't seen more development. Moving it as an "idea" in the discussions.

If you company is interested in the development of this feature please let me know.

**Comment by nkraemer2** [comment]
I encountered a similar need and was able to work around it [based on this SO post](https://stackoverflow.com/questions/79359145/refresh-a-connection-token-in-psycopg3-connectionpool), but consider this a +1 for the idea of having an official option to update credentials.

**Reply by dvarrazzo** [reply]
@nkraemer2 What do you think would be better for your use case:

- a `Pool.set_config()` callback so that an external agent may change the pool configuration, or
- a `Pool.get_config` init parameter as a callable that a pool may invoke every time it requires a connection string?

In the first case, when an external agent changes the connection parameter, would it be required to clear the pool and obtain new connections, or would the connections already in the pool still be valid until their configured expiration?

**Reply by nkraemer2** [reply]
I can't speak for others but in my particular use case:

- A Pool.get_config callback would be most convenient
- A Pool.set_config() would also work

In either case, I would not need existing connections to be cleared, I would only need new connections to have the new information. Basically, my password is a token that expires periodically, and I need a way to renew it when creating new connections.

**Reply by tominator95** [reply]
I am the author of the [SO post ](https://stackoverflow.com/questions/79359145/refresh-a-connection-token-in-psycopg3-connectionpool). Thanks for the follow-up @dvarrazzo!

Between the two options, I agree with @nkraemer2 and think the `get_config` init parameter (a callable the pool can invoke whenever it needs a connection string) would fit my use case better. The reason is that our connection tokens are short-lived (in my case 1 hour) and can expire or be rotated externally, so it's important that the pool always uses a fresh token when establishing new connections — without requiring manual intervention each time.

Regarding existing connections: if the token used by those is no longer valid (due to expiration or revocation), then yes, we would need to drop or recreate those connections. As a workaround, I've been using the max_lifetime setting (less than a hour) in the pool to ensure that connections eventually expire and get recreated with the new token.

Really appreciate the consideration of this feature — it would be super helpful!

**Comment by dvarrazzo** [comment]
I have created ticket #1042 for the feature (there doesn't seem to be a "convert to issue" feature in Github).

As mentioned in the issue, I expect to receive compensation for this development: either in the form of a project sponsorship or as a professional consultancy followed by a regular invoice if this comes easier for your company's accountancy.

**Comment by codeech** [comment]
FWIW, I subclassed AsyncConnection into my own `IAMTokenAsyncConnection` class to take care of this. No need to restart my service. I'm using AWS

```
class IAMTokenAsyncConnection(AsyncConnection):
    @classmethod
    async def connect(cls, conninfo: str = "", **kwargs) -> AsyncConnection:
        return await super().connect(
            your_custom_method_that_gets_new_conn_str_w_updated_credentials_if_needed(conninfo),
            **kwargs,
        )
```

```
        pool = AsyncConnectionPool(
            conninfo="some_connection_string",
            connection_class=IAMTokenAsyncConnection,
            min_size=1,
            max_size=10,
            max_lifetime=1800,
            open=False,  # Opening in __init__ is deprecated in psycopg_pool 3.2. See https://www.psycopg.org/psycopg3/docs/api/pool.html#the-asyncconnectionpool-class
        )
```

**Comment by epistoteles** [comment]
Weighting in here, I also have a need for rotating credentials in my AsyncConnectionPool. My host stays the same.

My preferred way to see this implemented would be to set `max_lifetime` to whatever the validity of the tokens is and have the option to pass a callable or coroutine as `conninfo` (instead of just a static string).

**Comment by SushanthK07** [comment]
I'm facing the same credentials rotation issue with **AWS RDS Proxy using IAM authentication**.

**My Problem:**
- AWS RDS IAM tokens expire every **15 minutes**
- After token expiry, new connections fail because psycopg3 pool reuses the original expired token
- I can't abruptly kill and recreate the pool because existing connections might be actively processing transactions
- I only need to update the **password field** (IAM token) - all other connection parameters stay the same

**Ideal Solution:**
Either the `Pool.get_config` callable approach mentioned above, or a password-specific callback can help

**Comment by dsuhinin** [comment]
Hey! What do you think about - https://github.com/psycopg/psycopg/pull/1167 ?

**Comment by dvarrazzo** [comment]
Heads up: thanks to @dsuhinin work this feature is being implemented in #1167. Long story short: the `conninfo` and `kwargs` parameters can now be callable too, which will be called at every connection attempt.

I would like to document the feature in [the pool page](https://www.psycopg.org/psycopg3/docs/advanced/pool.html). Does anyone have a simple practical example to contribute? For example to integrate the pool with one of the cloud providers.

**Reply by rcleveng** [reply]
I did something with Django to get this to work with Cloud SQL [https://github.com/rcleveng/django_gcp_iam_auth] (which I was using with Django now using pysopg3 without django) I've stopped using IAM Auth but will give this a try. Thank you so much for implementing it!

**Reply by rcleveng** [reply]
It worked perfectly, I wrote this up yesterday using it. https://github.com/rcleveng/simple_gcp_connector

**Comment by dvarrazzo** [comment]
Implemented in psycopg_pool 3.3

**Reply by ratkins** [reply]
Looking to upgrade right now, ~it doesn't seem to be mentioned in the release notes?~ Ah, maybe I'm not looking at the `_pool` docs.

**Reply by rcleveng** [reply]
Yes, it's in the pool docs.  I have an example for using it here too (if that helps)
https://github.com/rcleveng/simple_gcp_connector
